src/lsptoolshost/server/serverCommands.ts

```diff
@@ -0,0 +1,156 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { RoslynLanguageServer } from '../server/roslynLanguageServer';
+import { UriConverter } from '../utils/uriConverter';
+import * as languageClient from 'vscode-languageclient/node';
+
+/**
+ * Register server -> client commands as well as commands that drive the Roslyn Language Server.
+ */
+export function registerServerCommands(
+    context: vscode.ExtensionContext,
+    languageServer: RoslynLanguageServer,
+    outputChannel: vscode.LogOutputChannel
+) {
+    // It is very important to be careful about the types used as parameters for these command callbacks.
+    // If the arguments are coming from the server as json, it is NOT appropriate to use type definitions
+    // from the normal vscode API (e.g. vscode.Location) as input parameters.
+    //
+    // This is because at runtime the json objects do not contain the expected prototypes that the vscode types
+    // have and will fail 'instanceof' checks that are sprinkled throught the vscode APIs.
+    //
+    // Instead, we define inputs from the server using the LSP type definitions as those have no prototypes
+    // so we don't accidentally pass them directly into vscode APIs.
+    context.subscriptions.push(vscode.commands.registerCommand('roslyn.client.peekReferences', peekReferencesCallback));
+    context.subscriptions.push(
+        vscode.commands.registerCommand(
+            'roslyn.client.completionComplexEdit',
```

Comment:
It would be a fair argument that these server -> client commands could be organized beneath an appropriate feature folder.